pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,488 @@
+		return nil, fmt.Errorf("failed to find node from machine %q by ProviderID: %v", machine.GetName(), err)
+	}
+	if node != nil {
+		return node, nil
+	}
+
+	node, err = r.findNodeFromMachineByIP(machine)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find node from machine %q by internal IP: %v", machine.GetName(), err)
+	}
+	return node, nil
+}
+
+func (r *ReconcileNodeLink) findNodeFromMachineByProviderID(machine *mapiv1beta1.Machine) (*corev1.Node, error) {
+	if machine == nil {
+		klog.V(3).Infof("Finding node from machine nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding node from machine %q by providerID", machine.GetName())
+	if machine.Spec.ProviderID == nil {
+		klog.Warningf("Machine %q has no providerID", machine.GetName())
+		return nil, nil
+	}
+
+	nodes, err := r.listNodesByFieldFunc(nodeProviderIDIndex, *machine.Spec.ProviderID)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting node list: %v", err)
+	}
+
+	if len(nodes) > 0 {
+		klog.V(3).Infof("Found node %q for machine %q with providerID %q", nodes[0].GetName(), machine.GetName(), nodes[0].Spec.ProviderID)
+		return nodes[0].DeepCopy(), nil
+	}
+
+	return nil, nil
+}
+
+func (r *ReconcileNodeLink) findNodeFromMachineByIP(machine *mapiv1beta1.Machine) (*corev1.Node, error) {
+	if machine == nil {
+		klog.V(3).Infof("Finding node from machine nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding node from machine %q by IP", machine.GetName())
+	var machineInternalAddress string
+	for _, a := range machine.Status.Addresses {
+		if a.Type == corev1.NodeInternalIP {
+			machineInternalAddress = a.Address
+			klog.V(3).Infof("Found internal IP for machine %q: %q", machine.GetName(), machineInternalAddress)
+			break
+		}
+	}
+
+	if machineInternalAddress == "" {
+		klog.Warningf("not found internal IP for machine %q", machine.GetName())
+		return nil, nil
+	}
+
+	nodes, err := r.listNodesByFieldFunc(nodeInternalIPIndex, machineInternalAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting node list: %v", err)
+	}
+
+	if len(nodes) > 0 {
+		klog.V(3).Infof("Found node %q for machine %q with internal IP %q", nodes[0].GetName(), machine.GetName(), machineInternalAddress)
+		return nodes[0].DeepCopy(), nil
+	}
+
+	klog.V(3).Infof("not found node match for machine %q with internal IP %q", machine.GetName(), machineInternalAddress)
+	return nil, nil
+}
+
+func (r *ReconcileNodeLink) findMachineFromNode(node *corev1.Node) (*mapiv1beta1.Machine, error) {
+	if node == nil {
+		klog.V(3).Infof("Finding machine from node nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding machine from node %q", node.GetName())
+	machine, err := r.findMachineFromNodeByProviderID(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find machine from node %q by ProviderID: %v", node.GetName(), err)
+	}
+	if machine != nil {
+		return machine, nil
+	}
+
+	machine, err = r.findMachineFromNodeByIP(node)
```

Comment:
Should we set node.Spec.ProviderID from machine.Spec.ProviderID?

Reply:
> Should we set node.Spec.ProviderID from machine.Spec.ProviderID?

Not here - this function is about find.